app/src/main/java/com/android/sample/model/user/UserViewModel.kt

```diff
@@ -192,54 +196,94 @@ class UserViewModel(
   }
 
   /**
-   * Adds an ingredient to the user fridge. More precisely, it increases the count of said
-   * ingredient in the fridge.
+   * Method that add an [Ingredient] to the fridge of the user
    *
    * @param ingredient the ingredient to be added
-   * @param count the number of ingredients to be added
+   * @param quantity the quantity of the ingredient
+   * @param expirationDate the expiration date of the ingredient
    */
-  fun addIngredientToUserFridge(ingredient: Ingredient, count: Int = 1) {
-    try {
-      val changedElement: Pair<Ingredient, Int> =
-          _fridge.value.first { it.first.barCode == ingredient.barCode }
-
-      updateList(_fridge, changedElement, add = false)
-      updateList(_fridge, Pair(changedElement.first, changedElement.second + count), add = true)
-    } catch (e: NoSuchElementException) {
-      updateList(_fridge, Pair(ingredient, count), add = true)
-    }
-    updateCurrentUser()
+  private fun addIngredientToUserFridge(
+      ingredient: Ingredient,
+      quantity: String,
+      expirationDate: LocalDate
+  ) {
+    val newFridgeItem = FridgeItem(ingredient.barCode.toString(), quantity, expirationDate)
+    updateList(_fridgeItems, Pair(newFridgeItem, ingredient), add = true)
   }
 
   /**
-   * Removes an ingredient from the user fridge. More precisely, it will decrease the count of said
-   * ingredient in the fridge.
+   * Method that removes an [Ingredient] from the fridge of the user and updates the user in the
+   * database
    *
    * @param ingredient the ingredient to be removed
-   * @param count the number of ingredients to be removed
    * @throws IllegalArgumentException if the count is greater than the number of ingredients in the
    *   fridge
    */
-  fun removeIngredientFromUserFridge(ingredient: Ingredient, count: Int = 1) {
+  fun removeIngredientFromUserFridge(ingredient: Ingredient) {
     try {
-      val changedElement: Pair<Ingredient, Int> =
-          _fridge.value.first { it.first.barCode == ingredient.barCode }
-
-      if (changedElement.second - count > 0) {
-        updateList(_fridge, changedElement, add = false)
-        updateList(_fridge, Pair(changedElement.first, changedElement.second - count), add = true)
-      } else if (changedElement.second - count == 0) {
-        updateList(_fridge, changedElement, add = false)
-        updateList(_fridge, changedElement, add = false)
-      } else {
-        throw IllegalArgumentException(REMOVED_TOO_MANY_INGREDIENTS_ERROR)
-      }
+      val deletedIngredient: Pair<FridgeItem, Ingredient> =
+          _fridgeItems.value.first { it.first.id == ingredient.barCode.toString() }
```

Comment:
It seems that you are deleting an ingredient by only checking its ID. However, based on the rest of your implementation, it appears that the same ingredient can exist with different expiration dates (which is great, your handling of this in `updateIngredientFromFridge` is a nice feature, allowing users to track multiple entries of the same item with distinct expiration dates).

This approach is problematic because the first ingredient matching the barcode may not necessarily be the one the user intends to delete. To address this, the function should also accept the expiration date as a parameter and use it to identify the correct ingredient to delete. For example, you could incorporate logic similar to what you used in `updateIngredientFromFridge`:

```kotlin
val changedIngredient: Pair<FridgeItem, Ingredient> = 
    _fridgeItems.value.first { 
        (it.first.id == ingredient.barCode.toString()) && 
        (it.first.expirationDate == expirationDate) 
    }
```

This would ensure that the deletion targets the intended ingredient. What do you think?

Reply:
Good catch. I made the decision to add multiple times the same ingredient but with different expiration dates after the implementation of the deletion of an ingretient. Thus I forgot to update the method.
It is updated now.